### ISSUE TITLE
Custom errors do not work

### ISSUE THREAD

**Issue by Skully71** [MightyPirates/OpenComputers#2542]
As long as a custom error contains `code` you get the proper exit code. However if it does not contain any `code` then the exit code is `0` without any form of message.

```
local function test_error()
  local _exception_mt = { __tostring = function(e) return 'ERROR: ' .. e.msg end }
  local SomethingBad = { err = 121, msg = 'Oops' }
  setmetatable(SomethingBad, _exception_mt)
  error(SomethingBad)
end

test_error()
```

See also http://www.lua.org/wshop06/Belmonte.pdf "A better error object"

This is a direct result of returning `0` at https://github.com/MightyPirates/OpenComputers/blob/c4c3811376c2a694f8d0a7ab732f643ba9c149a7/src/main/resources/assets/opencomputers/loot/openos/lib/process.lua#L73 instead of continuing into the regular error handling.

**Comment by payonel** [comment]
Looks like you're asking for shell support for custom errors. The current behavior is not out of spec, I haven't defined what the shell should do in this scenario. If you want your code to print its error message to the shell, and have a proper exit code, then use the expected error format, i.e.
reason[string]: anything but "terminated" is printed to stderr
code[number]: exit code, 0 if undefined

It could be argued that the exit code should default to 1. I find this a reasonable change to consider.

It could be argued that the custom error object should be tostring() called if no `reason` is given. I find this less compelling, but your code sample will print "nil" in the current state of the code, which is less than ideal. I can see either calling tostring on the msg itself (which I believe you are asking), or just stderr print ("unknown error"). I'll give this some thought.

If you have further insights or concerns, let me know. Thanks

**Comment by payonel** [comment]
perhaps one thing to keep in mind is that the openos shell (prompt) is not a lua shell, but instead it is trying to be more linux shell like. errors are not necessarily expected to behave like your lua prompt would

**Comment by Skully71** [comment]
Is there a document somewhere that outlines the spec for error objects? I can't find one.

When I run the above through Lua 5.2 I get: `bin\lua52.exe: ERROR: Oops`
(no traceback, which is awkward)

https://github.com/MightyPirates/OpenComputers/blob/c4c3811376c2a694f8d0a7ab732f643ba9c149a7/src/main/resources/assets/opencomputers/loot/openos/lib/process.lua#L76 already results in a `__tostring` call.

Having a code, reason handling is a good thing, but that I believe is out of spec. At least there should be some documentation for it up on http://ocdoc.cil.li/.

Whenever an error does not adhere to the code, reason requirements, then OC should opt for default handling.

**Comment by payonel** [comment]
I like the points you are making, I'll give this some thought and review. Thanks

**Comment by payonel** [comment]
closed with 355a5209771d19f874c3a8f31573288e18e85fc2
